File: 程序员面试金典/5.3翻转数位/reverseBits.java
class Solution {
    public int reverseBits(int num) {
        if (num == -1)
            return 32;
        int currentLength = 0, prevLength = 0, maxLength = 1;
        while (num != 0) {
            if ((num & 1) == 1)
                ++currentLength;
            else {
                prevLength = (num & 2) == 0 ? 0 : currentLength;
                currentLength = 0;
            }
            maxLength = Math.max(maxLength, prevLength + currentLength + 1);
            num >>>= 1;
        }
        return maxLength;
    }
}

File: 程序员面试金典/5.3/reverseBits.py
class Solution:
    def reverseBits(self, num: int) -> int:
        counts = []
        count = 0
        while num:
            if num % 2 == 1:
                count += 1
            else:
                counts.append(count)
                count = 0
            num //= 2
        counts.append(count)
        return max(counts[i] + counts[i + 1] + 1 for i in range(len(counts) - 1)) if len(counts) > 1 else counts[0] + 1
